File: typescript/src/middleware/guardrails/nudge.ts
export enum NudgeKind {
  RETRY = "retry",
  UNKNOWN_TOOL = "unknown_tool",
  STEP = "step",
  PREREQUISITE = "prerequisite",
}

export interface Nudge {
  role: string;
  content: string;
  kind: NudgeKind;
  tier: number;
}

function joinToolNames(names: string[]): string {
  if (names.length === 0) {
    return "(no tools available)";
  }
  if (names.length === 1) {
    return names[0];
  }
  if (names.length === 2) {
    return `${names[0]} and ${names[1]}`;
  }
  return names.slice(0, -1).join(", ") + `, and ${names[names.length - 1]}`;
}

export function retryNudge(_rawResponse: string, toolNames: string[]): Nudge {
  const toolsList = joinToolNames(toolNames);
  return {
    role: "user",
    content: `Your previous response was not a valid tool call. You must respond with a tool call, not free text. Available tools: ${toolsList}. Please try again with a valid tool call.`,
    kind: NudgeKind.RETRY,
    tier: 0,
  };
}

export function unknownToolNudge(toolName: string, toolNames: string[]): Nudge {
  const toolsList = joinToolNames(toolNames);
  return {
    role: "user",
    content: `Tool '${toolName}' does not exist. Available tools: ${toolsList}. Call one of them.`,
    kind: NudgeKind.UNKNOWN_TOOL,
    tier: 0,
  };
}

export function stepNudge(
  terminalTool: string,
  pendingSteps: string[],
  tier: number = 1,
): Nudge {
  const clampedTier = Math.max(1, Math.min(3, tier));
  const steps = joinToolNames(pendingSteps);

  let content: string;
  switch (clampedTier) {
    case 1:
      content = `You cannot call ${terminalTool} yet. You must first complete these required steps: ${steps}. Call one of them now.`;
      break;
    case 2:
      content = `You must call one of these tools now: ${steps}. Pick one.`;
      break;
    default:
      content = `STOP. You MUST call one of: ${steps}. Do NOT call ${terminalTool}. Your next response MUST be a tool call to one of: ${steps}.`;
  }

  return {
    role: "user",
    content,
    kind: NudgeKind.STEP,
    tier: clampedTier,
  };
}

export function prerequisiteNudge(
  toolName: string,
  missingPrereqs: string[],
): Nudge {
  const prereqs = joinToolNames(missingPrereqs);
  return {
    role: "user",
    content: `You cannot call ${toolName} yet. You must first call: ${prereqs}. Call the prerequisite tool now.`,
    kind: NudgeKind.PREREQUISITE,
    tier: 0,
  };
}

File: go/middleware/guardrails/nudge.go
package guardrails

import "fmt"

type NudgeKind string

const (
	NudgeKindRetry        NudgeKind = "retry"
	NudgeKindUnknownTool  NudgeKind = "unknown_tool"
	NudgeKindStep         NudgeKind = "step"
	NudgeKindPrerequisite NudgeKind = "prerequisite"
)

type Nudge struct {
	Role    string
	Content string
	Kind    NudgeKind
	Tier    int
}

func RetryNudge(rawResponse string, toolNames []string) *Nudge {
	toolsList := joinToolNames(toolNames)
	return &Nudge{
		Role:    "user",
		Content: fmt.Sprintf("Your previous response was not a valid tool call. You must respond with a tool call, not free text. Available tools: %s. Please try again with a valid tool call.", toolsList),
		Kind:    NudgeKindRetry,
		Tier:    0,
	}
}

func UnknownToolNudge(toolName string, toolNames []string) *Nudge {
	toolsList := joinToolNames(toolNames)
	return &Nudge{
		Role:    "user",
		Content: fmt.Sprintf("Tool '%s' does not exist. Available tools: %s. Call one of them.", toolName, toolsList),
		Kind:    NudgeKindUnknownTool,
		Tier:    0,
	}
}

func StepNudge(terminalTool string, pendingSteps []string, tier int) *Nudge {
	if tier < 1 {
		tier = 1
	}
	if tier > 3 {
		tier = 3
	}

	steps := joinToolNames(pendingSteps)

	var content string
	switch tier {
	case 1:
		content = fmt.Sprintf("You cannot call %s yet. You must first complete these required steps: %s. Call one of them now.", terminalTool, steps)
	case 2:
		content = fmt.Sprintf("You must call one of these tools now: %s. Pick one.", steps)
	default:
		content = fmt.Sprintf("STOP. You MUST call one of: %s. Do NOT call %s. Your next response MUST be a tool call to one of: %s.", steps, terminalTool, steps)
	}

	return &Nudge{
		Role:    "user",
		Content: content,
		Kind:    NudgeKindStep,
		Tier:    tier,
	}
}

func PrerequisiteNudge(toolName string, missingPrereqs []string) *Nudge {
	prereqs := joinToolNames(missingPrereqs)
	return &Nudge{
		Role:    "user",
		Content: fmt.Sprintf("You cannot call %s yet. You must first call: %s. Call the prerequisite tool now.", toolName, prereqs),
		Kind:    NudgeKindPrerequisite,
		Tier:    0,
	}
}

func joinToolNames(names []string) string {
	if len(names) == 0 {
		return "(no tools available)"
	}
	if len(names) == 1 {
		return names[0]
	}
	if len(names) == 2 {
		return names[0] + " and " + names[1]
	}
	result := names[0]
	for i := 1; i < len(names); i++ {
		if i == len(names)-1 {
			result += ", and " + names[i]
		} else {
			result += ", " + names[i]
		}
	}
	return result
}
